Fix deprecation warnings for upcoming ROS Galactic release

jacobperron opened the pull request:
Building against the latest version of rclcpp produces some deprecation warnings. This PR fixes warnings related to `rclcpp::Duration` construction (6b36efd) and `rclcpp::Node::declare_parameter` (7055b82).

Related tickets upstream, deprecating usage:

- https://github.com/ros2/rclcpp/pull/1432
- https://github.com/ros2/rclcpp/pull/1522

The second commit (7055b82) related to declare_parameter is not compatible with Foxy, so I recommend creating a new `galactic-devel` branch and we can retarget this PR.

jacobperron committed:
Fix deprecation warnings when constructing rclcpp::Duration objects

If we are constructing from nanoseconds we must be explicit about the units.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Fix deprecated calls to declare_parameter

If a default value is not provided, then we must specify the parameter type.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

SteveMacenski commented on src/ros_filter.cpp:
```diff
@@ -834,18 +834,18 @@ void RosFilter<T>::loadParams()
   double offset_tmp = this->declare_parameter("transform_time_offset", 0.0);
   tf_time_offset_ =
-    rclcpp::Duration(filter_utilities::secToNanosec(offset_tmp));
+    rclcpp::Duration::from_nanoseconds(filter_utilities::secToNanosec(offset_tmp));
```
this and the 2 under are converting from s -> ns -> duration. You could remove the `filter_utilities::secToNanosec` call and replace it with just `from_seconds`

jacobperron replied:
Of course :upside_down_face: 
See ca9e807

SteveMacenski commented on src/ros_filter.cpp:
```diff
@@ -604,7 +604,7 @@ void RosFilter<T>::integrateMeasurements(const rclcpp::Time & current_time)
         first_measurement->topic_name_;
       // revertTo may invalidate first_measurement
-      if (!revertTo(first_measurement_time - rclcpp::Duration(1))) {
+      if (!revertTo(first_measurement_time - rclcpp::Duration(1ns))) {
```
@ayrton04 should this be 1ns, 1s, ...?

jacobperron replied:
Thanks, I meant to ask the same question.

jacobperron replied:
If I understand the equivalent ROS 1 code, it is also subtracting 1 nanosecond:

https://github.com/cra-ros-pkg/robot_localization/blob/a2da33bb3aaa4b10358de391221fe0ed97b16f49/src/ros_filter.cpp#L611

ayrton04 replied:
So the PRs involved here are

https://github.com/cra-ros-pkg/robot_localization/pull/266/files

and

https://github.com/cra-ros-pkg/robot_localization/pull/279/files

The only reason we subtract any time at all is to ensure that we go to a measurement (presumably the first) that is before `first_measurement_time`. But we could also just change this to use `>=`:

https://github.com/cra-ros-pkg/robot_localization/blob/3326e80a4169365b4570b5adcd1a63f3e37ab54f/src/ros_filter.cpp#L3322-L3323

If I had to guess (I didn't dig too far), the measurement times were likely stored as doubles originally, and the `1e-9` thing was just me being careful with floating point numbers. But ROS times have exact representations, so I think the 1e-9 could go, and we can change the `revertTo` reverse iteration to check for `>=` instead of just `>`.

ayrton04 replied:
However, that change could be outside of scope for this. It was meant to be 1 nanosecond, yes.

SteveMacenski commented:
@jacobperron I opened a `ros2` branch to target it to since those `from_XYZ`'s aren't in foxy

jacobperron committed:
Refactor

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron commented:
Hold up, there's more warnings in other files I've overlooked.

jacobperron committed:
Fix other deprecated uses of rclcpp::Duration

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron commented:
> Hold up, there's more warnings in other files I've overlooked.

Okay, I think I got them all.

SteveMacenski commented on src/robot_localization_estimator.cpp:
```diff
@@ -182,7 +182,7 @@ void RobotLocalizationEstimator::extrapolate(
     boundary_state.time_stamp *
     1000000000);
-  rclcpp::Duration delta_duration = rclcpp::Duration(delta * 1000000000);
+  rclcpp::Duration delta_duration = rclcpp::Duration::from_nanoseconds(delta * 1000000000);
```
`from_seconds()` and remove the 1e9 multiplier

jacobperron replied:
Done 774841b

SteveMacenski commented:
Otherwise LGTM, waiting for Tom's response on that one comment

jacobperron committed:
Minor refactor

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron commented on src/ros_filter.cpp:
```diff
@@ -815,7 +815,7 @@ void RosFilter<T>::loadParams()
   std::string tf_prefix = "";
   std::string tf_prefix_path = "";
-  this->declare_parameter("tf_prefix");
+  this->declare_parameter<std::string>("tf_prefix");
```
I misunderstood how the new static type enforcement works. This code will throw an exception at this line if "tf_prefix" is not provided, but we want it to be optional.

jacobperron replied:
I've addressed this by handling NoParameterOverrideProvided exceptions when no parameters are provided: 3326e80

jacobperron committed:
Catch exceptions when no parameter override is provided

When there is a NoParameterOverrideProvided exception, the parameter
is not declared so we don't need to undeclare it.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

ayrton04 commented:
@SteveMacenski, feel free to merge. I know you said LGTM, but your review still had comments requested, so I re-requested your review.
